src/kafka-util/src/admin.rs

```diff
@@ -0,0 +1,142 @@
+/// metadata.
+///
+/// This function is a wrapper around [`AdminClient::create_topics`] that
+/// attempts to ensure the topic creation has propagated throughout the Kafka
+/// cluster before returning. Kafka topic creation is asynchronous, so
+/// attempting to consume from or produce to a topic immediately after its
+/// creation can result in "unknown topic" errors.
+///
+/// This function does not return successfully unless it can find the metadata
+/// for the newly-created topic in a call to [`Client::fetch_metadata`] and
+/// verify that the metadata reports the topic has the number of partitions
+/// requested in `new_topic`. Empirically, this seems to be the condition that
+/// guarantees that future attempts to consume from or produce to the topic will
+/// succeed.
+pub async fn create_topic<'a, C>(
+    client: &'a AdminClient<C>,
+    admin_opts: &AdminOptions,
+    new_topic: &'a NewTopic<'a>,
+) -> Result<(), CreateTopicError>
+where
+    C: ClientContext,
+{
+    let res = client
+        .create_topics(iter::once(new_topic), &admin_opts)
+        .await?;
+    if res.len() != 1 {
+        return Err(CreateTopicError::InvariantViolation(format!(
+            "kafka topic creation returned {} results, but exactly one result was expected",
+            res.len()
+        )));
+    }
+    match res.into_element() {
+        Ok(_) | Err((_, RDKafkaError::TopicAlreadyExists)) => Ok(()),
+        Err((_, e)) => Err(CreateTopicError::Kafka(KafkaError::AdminOp(e))),
+    }?;
+
+    // Topic creation is asynchronous, and if we don't wait for it to
+    // complete, we might produce a message (below) that causes it to
+    // get automatically created with multiple partitions. (Since
+    // multiple partitions have no ordering guarantees, this violates
+    // many assumptions that our tests make.)
+    retry::retry_for(Duration::from_secs(8), |_| async {
+        let metadata = client
+            .inner()
+            // N.B. It is extremely important not to ask specifically
+            // about the topic here, even though the API supports it!
+            // Asking about the topic will create it automatically...
+            // with the wrong number of partitions. Yes, this is
+            // unbelievably horrible.
+            .fetch_metadata(None, Some(Duration::from_secs(1)))?;
+        let topic = metadata
+            .topics()
+            .iter()
+            .find(|t| t.name() == new_topic.name)
+            .ok_or(CreateTopicError::MissingMetadata)?;
+        if topic.partitions().len() as i32 != new_topic.num_partitions {
+            return Err(CreateTopicError::PartitionCountMismatch {
+                expected: new_topic.num_partitions,
+                actual: topic.partitions().len() as i32,
+            });
+        }
+        Ok(())
+    })
+    .await
+}
+
+/// An error while creating a Kafka topic.
+#[derive(Debug)]
+pub enum CreateTopicError {
```

Comment:
this is almost certainly a dumb q but what is the benefit of having a structured error enum like this instead of just `bail`-ing with strings? my understanding is that it allows the caller to more easily program different actions for different kinds of errors -- is that basically it or is there more to it?

Reply:
Yup, exactly. I did it like this purely because testdrive doesn't use anyhow. It probably should, and if it learns to then we could just bail here and things would get a lot simpler.